File: python/cote/LTCD/전화번호_문자_조합.py
from typing import List
import random

N = 10
TEST_CASES = 10

dic = {
	"2" : "abc",
	"3" : "def",
	"4" : "ghi",
	"5" : "jkl",
	"6" : "mno",
	"7" : "pqks",
	"8" : "tuv",
	"9" : "wxyz"
}

def solve(digits: str) -> List[str] :
	global dic
		
	ret = []

	def dfs(i: int, acc: str) :
		if i == len(digits) :
			ret.append(acc)
			return


		for j in dic[digits[i]] : 
			dfs(i+1, acc + j)

		return

	dfs(0, "")

	return ret

if __name__ == "__main__" :
	random.seed(43)
	for _ in range(TEST_CASES) :
		digits = "".join([str(random.randrange(2,9)) for _ in range(random.randrange(1, N))])
		print(solve(digits))

File: javascript/cote/LTCD/전화번호_문자_조합.mjs

const N = 10
const TEST_CASES = 10

const dics = new Map(Object.entries({
	"2" : "abc",
	"3" : "def",
	"4" : "ghi",
	"5" : "jkl",
	"6" : "mno",
	"7" : "pqks",
	"8" : "tuv",
	"9" : "wxyz"
}))

function solve(digits) {

	const l = [...digits]
	const [stack, ret] = [[],[]]

	function dfs() {
		if (l.length == 0) {
			ret.push([...stack].join(""))
			return
		}

		const cur = l.shift() 
		for (let c of dics.get(cur)) {
			stack.push(c)
			dfs()
			stack.pop()
		}
		l.unshift(cur)

	}
	
	dfs()

	return ret
}


function main() {

	for (let it = 0; it < TEST_CASES; it++) {
		const n = Math.floor(Math.random() * (N - 1)) + 1
		const digits = new Array(n).fill(0).map(_ => (Math.floor(Math.random() * 8) + 2).toString())
		console.log(solve(digits))
	}
}

main()